text/0112-server-runtime.md

```diff
@@ -0,0 +1,331 @@
+
+- `@lwc/engine-core`: Core logic shared by the different runtimes including the rendering engine and the reactivity mechanism. This package should never be consumed directly in an application. This package is agnostic on the underlying rendering medium. It only provides APIs for building custom runtimes.
+- `@lwc/engine-dom`: Runtime that can be used to render LWC component trees in a DOM environment. This package is built on top of `@lwc/engine-core`.
+- `@lwc/engine-server`: Runtime used that can be used to render LWC component trees as strings. This package is built on top of `@lwc/engine-core`.
+
+The existing `lwc` package stays untouched and will be used to distribute the different versions of the engine. From the developer perspective, the experience should be identical.
+
+**`c/app/app.js`:**
+
+```js
+import { LightningElement } from "lwc";
+
+export default class App extends LightningElement {}
+```
+
+**`client.js`:**
+
+```js
+import { createElement } from "lwc"; // Aliased to @lwc/engine-dom
+import App from "c/app";
+
+const app = createElement("c-app", { is: App });
+document.body.appendChild(app);
+```
+
+**`server.js`:**
+
+```js
+import { createElement, renderToString } from "lwc"; // Aliased to @lwc/engine-server
+import App from "c/app";
+
+const app = createElement("c-app", { is: App });
+const str = renderToString(app);
+
+console.log(str);
+```
+
+### `@lwc/engine-core`
+
+This packages exposes the following platform-agnostic APIs:
+
+- `LightningElement`
+- `api`
+- `track`
+- `readonly`
+- `wire`
+- `setFeatureFlag`
+- `getComponentDef`
+- `isComponentConstructor`
+- `getComponentConstructor`
+- `unwrap`
+- `registerTemplate`
+- `registerComponent`
+- `registerDecorators`
+- `sanitizeAttribute`
+
+The DOM APIs used by the rendered engine are injected by the runtime depending on the environment. A list of all the DOM APIs the engine depends upon can be found in the [DOM APIs usage](#dom-apis-usage) section in the Appendix.
+
+### `@lwc/engine-dom`
+
+This package exposes the following APIs:
+
+- `createElement` + reaction hooks
+- `buildCustomElementConstructor`
+- `isNodeFromTemplate`
+
+This package injects the native DOM APIs into the `@lwc/engine-core` rendering engine.
+
+### `@lwc/engine-server`
+
+This package exposes the following APIs:
+
+- `createElement(name: string, options: { is: typeof LightningElement }): ServerHTMLElement`: This method creates a new LWC component tree. It follows the same signature as the `createElement` API from `@lwc/engine-dom`. Instead of returning a native `HTMLElement`, this method returns a `ServerHTMLElement` with the public properties, aria reflected properties and HTML global attributed.
+- `renderToString(element: ServerHTMLElement): string`: This method creates an LWC component tree synchronously and serialize it to string. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.
```

Comment:
One thing worth exploring: Can we pass `@api` data for the top level component here? If values are known at "tree creation time", we can probably hydrate a good portion of the LWC tree sync as the initial tree is being created

Reply:
this is probably another area to explore in a different RFC since it touches on how much can you render on the server synchronously.

Reply:
If the data is known before the component is rendered, the current proposal should be sufficient.

```js
import { createElement, renderToString } from "lwc";
import App from "c/app";

const app = createElement("c-app", { is: App });
app.foo = 'Some value';
app.bar = ['Some', 'other', 'value.'];
const str = renderToString(app);

console.log(str);
```

Like a standard `HTMLElement` returned by `createElement` when LWC is running in the browser, the `ServerHTMLElement` return by `created `createElement` on the server would expose:
- the component public properties
- the global HTML properties 
- aria reflection properties